### PR TITLE
Add the major version of the release as suffix to the package name

### PR DESCRIPTION
**Pull request by andreas-bok-sociomantic**
Addresses problem where apt-get install a newer version over an older. 
The PR adds the major version as suffix to the libname. 
Addresses #149

### CODE REVIEW

**Comment by mathias-lang-sociomantic** [comment]
Squash ?

**Comment by andreas-bok-sociomantic** [comment]
Yes, but still a WIP, so will pingpong until the changes are more stable

File: deb/build [hunk 1]
```diff
@@ -22,6 +22,38 @@ genchangelog()
...
+    pkgname=$pkgname$pkgsuffix
+
+    # the versioned library name
+    so_libname=$pkgname.so
+
+    # create symbolic links for the renamed library
+    ln -sf $so_libname.$libversion $so_libname.$lib_major
+    ln -sf $so_libname.$lib_major $so_libname
+
+    cd $curr_dir
```
Comment (leandro-lucarella-sociomantic) [inline deb/build]: No need to save this, you can access the previous directory with `$OLDPWD` and you can just use `cd -` as a shortcut to `cd $OLDPWD`.

File: deb/build [hunk 2]
```diff
@@ -22,6 +22,38 @@ genchangelog()
...
+
+    curr_dir=$PWD
+
+    # cd to the `make` install path
+    cd $2
+
+    # rename the original library file to the versioned library file
+    mv $pkgname.so.$libversion "$pkgname$pkgsuffix.so.$libversion"
+    mv $pkgname.a "$pkgname$pkgsuffix.a"
+    mv $pkgname.la "$pkgname$pkgsuffix.la"
```
Comment (leandro-lucarella-sociomantic) [inline deb/build]: Why not doing all this when creating the package via `fpm` mappings instead? It looks overly complicated.

File: m4/ax_package_versioned.m4 [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# ===========================================================================
+# https://github.com/andreas-bok-sociomantic
```
Comment (leandro-lucarella-sociomantic) [inline m4/ax_package_versioned.m4]: Why not using the upstream project if you really need to specify an URL?

File: m4/ax_package_versioned.m4 [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#   and this notice are preserved. This file is offered as-is, without any
+#   warranty.
+
+#serial 3
+AC_DEFUN([AX_PACKAGE_VERSIONED],
+    [AC_PREREQ([2.63])dnl
+
+    major=`echo $VERSION | cut -f 1 -d .`
+    AC_SUBST([PACKAGE_VERSIONED],["$PACKAGE$major"])
+    ])
```
Comment (leandro-lucarella-sociomantic) [inline m4/ax_package_versioned.m4]: Also do you really need a m4 extension for this? Why not doing it in the `deb/build` script?

**Comment by andreas-bok-sociomantic** [comment]
Restructured based on comments from luca

**Comment by andreas-bok-sociomantic** [comment]
> I don't think the headers should include it, for that it might be better to install headers in a different directory, i.e.: /usr/include/libdrizzle-redux/libdrizzle-redux6/libdrizzle-redux so then an application could use -I/usr/include/libdrizzle-redux/libdrizzle-redux6

The changes doesn't yet include the installation of include into a sub directory of `/usr/include/libdrizzle-redux/` as suggested by @leandro-lucarella-sociomantic in [issue 149](https://github.com/sociomantic-tsunami/libdrizzle-redux/issues/149#issuecomment-295786481)

**Comment by andreas-bok-sociomantic** [comment]
Updated: 
- Fixed wrong declaration of `dist_pkgname`
- Added the mapping for the include headers to 

    `/usr/include/libdrizzle-redux/libdrizzle-redux6/libdrizzle-redux` 
```diff
diff --git a/deb/build b/deb/build
index 614a731..1270c0c 100755
--- a/deb/build
+++ b/deb/build
@@ -44,6 +44,12 @@ pkgname=libdrizzle-redux
 # library .so name
 so_libname=$pkgname.so
 
+# get the api level for the package from the cmd args
+api_level=""
+if [ $# -eq 1 -a ! -z "$1" ]; then
+    api_level=$1
+fi
+
 # the package name used in the distribution
 dist_pkgname=$pkgname$api_level
 
@@ -63,12 +69,6 @@ fi
 libversion=`echo $version_number | grep -o -P -e "([0-9]+\.?)+$"`
 lib_major=$(echo $libversion | cut -f1 -d .)
 
-# get the api level for the package from the cmd args
-api_level=""
-if [ $# -eq 1 -a ! -z "$1" ]; then
-    api_level=$1
-fi
-
 # create temp file to write changelog to
 changelog=`mktemp`
 trap "rm -f '$changelog'; exit 1" INT TERM QUIT
@@ -140,6 +140,6 @@ fpm -s dir -t deb -n "$dist_pkgname-dev" -v "$pkgversion" \
     --depends $dist_pkgname \
     --deb-changelog "$changelog" \
     --deb-no-default-config-files \
-    install/usr/include/=/usr/ \
+    install/usr/include/include/=/usr/include/$pkgname/$dist_pkgname/ \
     $libdir/$dist_so_libname=/usr/lib/ \
     $libdir/$pkgname.a=/usr/lib/$dist_pkgname.a

```

File: deb/build [hunk 5]
```diff
@@ -44,15 +69,29 @@ fi
...
+# create temp file to write changelog to
 changelog=`mktemp`
 trap "rm -f '$changelog'; exit 1" INT TERM QUIT
 
-so_libname="./install/usr/lib/$so_filename"
+# (re)create symbolic links with the distributed package name, dist_pkgname
+cd $libdir
+touch $dist_so_libname.$libversion
+ln -sf $dist_so_libname.$libversion $dist_so_libname.$lib_major
+ln -sf $dist_so_libname.$lib_major $dist_so_libname
```
Comment (ben-palmer-sociomantic) [inline deb/build:80]: So this line will replace the current `libdrizzle-redux` with `libdrizzle-redux.6` right? And that is intended?

File: deb/build [hunk 6]
```diff
@@ -66,41 +105,41 @@ fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
...
     --license 'Simplified BSD License' \
     --category libdevel \
-    --depends libdrizzle-redux \
+    --depends $dist_pkgname \
     --deb-changelog "$changelog" \
     --deb-no-default-config-files \
-    install/usr/include/=/usr/ \
-    $so_libname=/usr/lib/ \
-    install/usr/lib/libdrizzle-redux.a=/usr/lib/
+    install/usr/include/include/=/usr/include/$pkgname/$dist_pkgname/ \
```
Comment (ben-palmer-sociomantic) [inline deb/build:143]: Is `include/include` correct?

**Review by ben-palmer-sociomantic — COMMENTED** [review]
I had a couple of queries but generally LGTM.

File: deb/build (hunk 6 above)
Comment (andreas-bok-sociomantic) [reply deb/build:143]: Yes, with the new project layout the make install create a two `include` subdirs.

File: deb/build (hunk 5 above)
Comment (andreas-bok-sociomantic) [reply deb/build:80]: No the api level is appended right after the pkgname, ie. `api_level=6` would generate the package  `libdrizzle-redux6`

Comment (ben-palmer-sociomantic) [reply deb/build:80]: Ah ok, but this is the intended behaviour? Just wanted to confirm I understand whats going on.

Comment (andreas-bok-sociomantic) [reply deb/build:80]: Yeah. It would change the linking to -ldrizzle--redux6 for versions major branch 6.x.x.  Some libraries have use major-minor to distinguish the api level of the library, e.g. `somelib-2.0.so.1.0.1` but given that semver doesn't break ABI compatibility I reckon major version is sufficient.

**Comment by andreas-bok-sociomantic** [comment]
Updated 
- Specify api level for RPM packages

```diff 
diff --git a/configure.ac b/configure.ac
index 63ec798..b19f3d4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -16,6 +16,12 @@ AC_CANONICAL_TARGET
 AM_INIT_AUTOMAKE(1.11 no-define color-tests -Wno-portability subdir-objects foreign tar-ustar m4_ifndef([AM_WITH_REGEX], [serial-tests]))
 AC_PREREQ([2.68])
 
+# Define the api level of the package.
+AC_SUBST([API_LEVEL],[6])
+
+# Add the api level to the name of the generated tar file
+AC_SUBST([PACKAGE_TARNAME],[$PACKAGE_NAME$API_LEVEL])
+
 AC_CANONICAL_HOST
 AC_CANONICAL_BUILD
 AC_ARG_PROGRAM
@@ -144,8 +150,6 @@ AX_APPEND_COMPILE_FLAGS([-I./include], [CFLAGS])
 AX_CREATE_GENERIC_CONFIG
 AX_AM_JOBSERVER([yes])
 
-AC_SUBST([API_LEVEL],[6])
-
 AC_CONFIG_FILES([Makefile
                  docs/conf.py
                  rpm/spec
diff --git a/rpm/include.am b/rpm/include.am
index 4e76906..7b78dc1 100644
--- a/rpm/include.am
+++ b/rpm/include.am
@@ -2,8 +2,8 @@
 
 rpm-build: rpm/spec dist
 	@rm -f *.rpm
-	@rm -f ~/rpmbuild/RPMS/x86_64/$(PACKAGE)-$(VERSION)*.rpm
-	@rm -f ~/rpmbuild/SRPMS/$(PACKAGE)-$(VERSION)*.rpm
+	@rm -f ~/rpmbuild/RPMS/x86_64/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm
+	@rm -f ~/rpmbuild/SRPMS/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm
 	@mkdir -p ~/rpmbuild/BUILD/
 	@mkdir -p ~/rpmbuild/RPMS/i386/
 	@mkdir -p ~/rpmbuild/RPMS/i686/
@@ -12,11 +12,11 @@ rpm-build: rpm/spec dist
 	@mkdir -p ~/rpmbuild/SOURCES/
 	@mkdir -p ~/rpmbuild/SPECS/
 	@mkdir -p ~/rpmbuild/SRPMS/
-	@cp $(PACKAGE)-$(VERSION).tar.gz ~/rpmbuild/SOURCES/
+	@cp $(PACKAGE)$(API_LEVEL)-$(VERSION).tar.gz ~/rpmbuild/SOURCES/
 	@rpmbuild -ba --clean rpm/spec
-	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)-$(VERSION)*.rpm .
-	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)-devel-$(VERSION)*.rpm .
-	@cp ~/rpmbuild/SRPMS/$(PACKAGE)-$(VERSION)*.rpm .
+	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm .
+	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)$(API_LEVEL)-devel-$(VERSION)*.rpm .
+	@cp ~/rpmbuild/SRPMS/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm .
 
 rpm-sign: rpm-build
 	@rpm --addsign *.rpm
diff --git a/rpm/spec.in b/rpm/spec.in
index 71672cc..f17ec79 100644
--- a/rpm/spec.in
+++ b/rpm/spec.in
@@ -1,5 +1,5 @@
 Summary: Simplified API to MySQL databases
-Name: @PACKAGE@
+Name: @PACKAGE@@API_LEVEL@
 Version: @VERSION@
 Release: 1
 License: BSD
@@ -11,7 +11,7 @@ URL: https://github.com/sociomantic-tsunami/libdrizzle-redux
 %define packager_email %(git config user.email)
 Packager: %{packager_name} <%{packager_email}>
 
-Source: libdrizzle-redux-%{version}.tar.gz
+Source: %{name}-%{version}.tar.gz
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-buildroot
 
 %description

```

**Comment by ben-palmer-sociomantic** [comment]
LGTM. Do you want any more reviews or are you happy for me to rebase?

**Comment by andreas-bok-sociomantic** [comment]
> LGTM. Do you want any more reviews or are you happy for me to rebase?

I might just wanna see if @leandro-lucarella-sociomantic has any comments. He after all has a `deb-conf` T-Shirt :)

**Comment by leandro-lucarella-sociomantic** [comment]
> He after all has a deb-conf

It is *DebConf* and I was just visiting some friends :P

File: deb/build [hunk 7]
```diff
@@ -2,6 +2,14 @@
...
 # script using fpm <https://github.com/jordansissel/fpm> to build libdrizzle-redux
 # runtime, debugging and development deb packages
+#
+# Usage: ./build [api_level]
+#
+#     api_level        the api level for the package
+#
+# The `api_level` will be appended to the `package_name` for the generated
+# packages, i.e. running `./build liblicious 6` would generate a packages named
+# `liblicious6`, `liblicious6-dbg`, `liblicious6-dev`.
```
Comment (leandro-lucarella-sociomantic) [inline deb/build]: Why liblicious?

File: deb/build [hunk 8]
```diff
@@ -32,10 +40,27 @@ pkgmaint=$(echo "`git config user.name` <`git config user.email`>")
...
 
+# the so library name used in the distribution
+dist_so_libname=$dist_pkgname.so
+
+# relative path to the directory with library files to install
+libdir="./install/usr/lib"
+
+# get the ABI library version from the generated .so file
+version_number=`find $libdir ! -type l -name "$so_libname.*"`
 if [ $(echo $version_number | wc -l) != 1 ]; then
```
Comment (leandro-lucarella-sociomantic) [inline deb/build]: Not part of this PR, but this is a very weird way to check if a file is found or not :thinking:

**Review by leandro-lucarella-sociomantic — COMMENTED** [review]
I did just a very quick pass, honestly I'm not an expert in regards of shared library versioning or autotools :/

File: deb/build (hunk 8 above)
Comment (andreas-bok-sociomantic) [reply deb/build]: Hehe, saying that `-e` is easier...

File: deb/build (hunk 7 above)
Comment (andreas-bok-sociomantic) [reply deb/build]: Bad joke, sorry

File: deb/build (hunk 8 above)
Comment (andreas-bok-sociomantic) [reply deb/build]: It could meaningfully go into the commit "Sanitize deb/build script"

**Comment by andreas-bok-sociomantic** [comment]
Updated based on comments from @leandro-lucarella-sociomantic 

```diff
diff --git a/Makefile.am b/Makefile.am
index 89864fe..5a274c9 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -11,6 +11,7 @@ CLEANFILES =
 DISTCLEANFILES =
 bin_PROGRAMS =
 dist_bin_SCRIPTS =
+distdir = @PACKAGE@@API_LEVEL@-@VERSION@
 noinst_HEADERS =
 lib_LTLIBRARIES =
 man_MANS =
@@ -69,8 +70,8 @@ maintainer-clean-local:
 	-rm m4/ltversion.m4
 	-rm m4/lt~obsolete.m4
 	find . -type f -name '*~' -exec rm -f '{}' \;
-	-rm -f @PACKAGE@-*.tar.gz
-	-rm -f rpm/@PACKAGE@-*.rpm
-	-rm -f @PACKAGE@-*.deb
+	-rm -f @PACKAGE@@API_LEVEL@-*.tar.gz
+	-rm -f rpm/@PACKAGE@@API_LEVEL@-*.rpm
+	-rm -f @PACKAGE@@API_LEVEL@-*.deb
 
 dist_bin_SCRIPTS+= @GENERIC_CONFIG@
diff --git a/deb/build b/deb/build
index 1270c0c..5ccb01b 100755
--- a/deb/build
+++ b/deb/build
@@ -8,8 +8,8 @@
 #     api_level        the api level for the package
 #
 # The `api_level` will be appended to the `package_name` for the generated
-# packages, i.e. running `./build liblicious 6` would generate a packages named
-# `liblicious6`, `liblicious6-dbg`, `liblicious6-dev`.
+# packages, i.e. running `./build 6` would generate a packages named
+# `libdrizzle-redux6`, `libdrizzle-redux-dbg`, `libdrizzle-redux-dev`.
 
 set -e
 
@@ -61,7 +61,7 @@ libdir="./install/usr/lib"
 
 # get the ABI library version from the generated .so file
 version_number=`find $libdir ! -type l -name "$so_libname.*"`
-if [ $(echo $version_number | wc -l) != 1 ]; then
+if [ ! -e $version_number ]; then
     echo "'.so' file not found"
     exit
```

File: deb/build [hunk 9]
```diff
@@ -1,8 +1,14 @@
 #!/bin/sh
+#
+# script using fpm <https://github.com/jordansissel/fpm> to build libdrizzle-redux
+# runtime, debugging and development deb packages
+
 set -e
 
+# change directory to containing deb `build` script
```
Comment (ben-palmer-sociomantic) [inline deb/build]: Perhaps better as `change directory to the deb `build` script directory`

File: deb/build [hunk 10]
```diff
@@ -2,6 +2,14 @@
...
 # script using fpm <https://github.com/jordansissel/fpm> to build libdrizzle-redux
 # runtime, debugging and development deb packages
+#
+# Usage: ./build [api_level]
+#
+#     api_level        the api level for the package
+#
+# The `api_level` will be appended to the `package_name` for the generated
+# packages, i.e. running `./build 6` would generate a packages named
+# `libdrizzle-redux6`, `libdrizzle-redux-dbg`, `libdrizzle-redux-dev`.
```
Comment (ben-palmer-sociomantic) [inline deb/build]: So `libdrizzle-redux6` is the only lib with the suffix? Not `libdrizzle-redux-dbg` and `libdrizzle-redux-dev`?

**Review by ben-palmer-sociomantic — COMMENTED** [review]
Couple of small comments then LGTM.

**Comment by andreas-bok-sociomantic** [comment]
Updated based on comments from @ben-palmer-sociomantic 
```diff
diff --git a/deb/build b/deb/build
index a117446..a0c909a 100755
--- a/deb/build
+++ b/deb/build
@@ -9,11 +9,11 @@
 #
 # The `api_level` will be appended to the `package_name` for the generated
 # packages, i.e. running `./build 6` would generate a packages named
-# `libdrizzle-redux6`, `libdrizzle-redux-dbg`, `libdrizzle-redux-dev`.
+# `libdrizzle-redux6`, `libdrizzle-redux6-dbg`, `libdrizzle-redux6-dev`.
 
 set -e
 
-# change directory to containing deb `build` script
+# cd to the directory containing the deb `build` script
 cd `dirname $0`
 
 # function to automatically generate changelog
```

**Comment by andreas-bok-sociomantic** [comment]
Updated
- Rebased on v6.0.x
- Fixed typo
- Updated sections about linking in README.rst and compiling.rst

File: README.rst [hunk 11]
```diff
@@ -71,10 +71,19 @@ refer to `compiling.rst`_ or build the documentation, cf. below.
 
 **Linking libdrizzle-redux**::
 
+Ensure the library is in your library and include paths. For releases prior to
+`v6.0.3` linking your app against libdrizzle-redux requires ``-ldrizzle-redux``::
```
Comment (ben-palmer-sociomantic) [inline README.rst]: `v6.0.3` or `v6.0.2`?

**Comment by ben-palmer-sociomantic** [comment]
One comment, then LGTM.

File: README.rst (hunk 11 above)
Comment (andreas-bok-sociomantic) [reply README.rst]: Ah, damn it. `v6.0.2`

**Comment by andreas-bok-sociomantic** [comment]
Fixed wrong version number in last commit

**Comment by ben-palmer-sociomantic** [comment]
LGTM

**Comment by andreas-bok-sociomantic** [comment]
Tx
